packs/newrelic/sensors/newrelic_application_sensor.py

```diff
@@ -0,0 +1,311 @@
+
+        elif self._is_alert_closed(long_description) or \
+             self._is_downtime_recovered(long_description):
+
+            payload = {
+                'alert': alert_body,
+                'header': hook_headers
+            }
+            eventlet.spawn_after(self._normal_report_delay, self._dispatch_server_normal,
+                                 payload)
+
+        elif self._is_alert_canceled(long_description) or \
+             self._is_alert_acknowledged(long_description):
+            self._log.info('Ignored alert : %s.', alert_body)
+
+    def _dispatch_server_normal(self, payload, attempt_no=0):
+        '''
+        Dispatches SERVER_NORMAL_TRIGGER_REF if the all servers health_status is 'green'.
+        '''
+        # basic guard to avoid queuing up forever.
+        if attempt_no == 10:
+            self._log.warn('Abandoning SERVER_NORMAL_TRIGGER_REF dispatch. Payload %s', payload)
+        servers = self._get_servers(payload['alert']['servers'])
+        # make sure all servers are ok.
+        all_servers_ok = True
+        for server in servers:
+            all_servers_ok &= server['health_status'] in ['green']
+            if not all_servers_ok:
+                break
+
+        if all_servers_ok:
+            self._sensor_service.dispatch(WEB_APP_NORMAL_TRIGGER_REF, payload)
+        else:
+            for server in servers:
+                self._log.info('server %s has state %s. Rescheduling normal check.',
+                               server['name'], server['health_status'])
+            eventlet.spawn_after(self._normal_report_delay, self._dispatch_server_normal,
+                                 payload, attempt_no + 1)
+
+    # alert test methods
+    def _is_alert_opened(self, long_description):
+        return long_description and long_description.startswith('Alert opened')
+
+    def _is_alert_closed(self, long_description):
+        return long_description and long_description.startswith('Alert ended')
+
+    def _is_alert_canceled(self, long_description):
+        return long_description and long_description.startswith('Alert canceled')
+
+    def _is_alert_acknowledged(self, long_description):
+        return long_description and long_description.startswith('Alert acknowledged')
+
+    def _is_escalated_downtime(self, long_description):
+        return long_description and long_description.startswith('Alert escalated to downtime')
+
+    def _is_downtime_recovered(self, long_description):
+        return long_description and long_description.startswith('Alert downtime recovered')
+
+    # newrelic API methods
+    def _get_application(self, app_name):
+        params = None
+        if app_name:
+            params = {'filter[name]': app_name}
+        url = urllib_parse.urljoin(self._api_url, 'applications.json')
+        resp = requests.get(url, headers=self._headers, params=params).json()
+        if 'applications' in resp:
```

Comment:
Ah, ignore my last comment.

This is used in a different context than I originally thought.